lib/argon2/password.rb

```diff
@@ -0,0 +1,191 @@
+module Argon2
+  ##
+  # Front-end API for the Argon2 module.
+  #
+  # TODO: Find a good place to document the following knowledge, it's a PITA to
+  #       google atm.
+  #
+  # 1) Variant used
+  # 2) Version used
+  # 3) Config used
+  #   a) m - m_cost (memory)
+  #   b) t - t_cost (time/stretches)
+  #   c) p - p_cost (parallelism/cores)
+  # 4) Salt
+  # 5) Checksum
+  #
+  # |   1   | 2  |       3       |           4           |      5    (truncated)
+  # $argon2i$v=19$m=65536,t=2,p=1$VG9vTG9uZ1NhbGVMZW5ndGg$mYleBHsG6N0+H4JGJ0xXoI
+  #
+  class Password
+    #
+    DEFAULT_T_COST = 2
+    DEFAULT_M_COST = 16
+    #
+    MIN_T_COST = 1
+    MAX_T_COST = 750
+    #
+    MIN_M_COST = 1
+    MAX_M_COST = 31
+    # The complete Argon2 digest string (not to be confused with the checksum).
+    attr_reader :digest
+    # The hash portion of the stored password hash.
+    attr_reader :checksum
+    # The salt of the store password hash (including version and cost).
+    attr_reader :salt
+    # Variant used (argon2[i|d|id])
+    attr_reader :variant
+    # The version of the argon2 algorithm used to create the hash.
+    attr_reader :version
+    # The time cost factor used to create the hash.
+    attr_reader :t_cost
+    # The memory cost factor used to create the hash.
+    attr_reader :m_cost
+    # The parallelism cost factor used to create the hash.
+    attr_reader :p_cost
+
+    class << self
+      def create(password, options = {})
+        raise Argon2::Errors::InvalidPassword unless password.is_a?(String)
+
+        t_cost = options[:t_cost] || DEFAULT_T_COST
+        m_cost = options[:m_cost] || DEFAULT_M_COST
+
+        if t_cost < MIN_T_COST || t_cost > MAX_T_COST
+          raise Argon2::Errors::InvalidTCost
```

Comment:
Just noticed that this is a bit conflicted with how InvalidTCost/etc inherit atm. InvalidCost is nested under InvalidHash due to being used in split_hash for reading the digest, but with its usage here, it would be better suited going under InvalidPassword.

Might need to reconsider the hierarchy and/or use slightly different naming and have two types of cost errors.

Reply:
I updated the documentation to mention that it gets used for two different scenarios. If anyone wants to split the two scenarios into two different Errors though, I'd be willing to take another look.